### DISCUSSION TITLE
Use vim commands (not user defined) to lazy-load using cmd

### DISCUSSION THREAD

**Discussion by ichigo-gyunyu** [wbthomason/packer.nvim#476] [Q&A] [answered]
Hello, is it possible to lazy load using vim commands? When I try to add `cmd = "write"` to one of my plugins to lazy-load, I get an error during `PackerCompile`
```
Vim(command):E183: User defined commands must start with an uppercase letter
Please check your config for correctness
```

**Comment by wbthomason** [comment, marked as answer]
In general, no; in this case, kind of.

In general, we need to be able to make a "fake" command or rely on the `CmdUndefined` event happening in order to lazy-load on a command. For built-in commands, as you've discovered, this is not possible (as far as I'm aware).

However, in the specific case of `:write` (and a few other built-in commands), you can instead lazy-load on the event (e.g. `BufWrite`) that will happen when the command is run, thereby indirectly lazy-loading on the command.

**Reply by ichigo-gyunyu** [reply]
Ah, I see. The "fake" command method seems interesting.

I am trying to lazy-load an autosave plugin, so I guess `BufWrite` would not be triggered until I manually save the first time. Oh well, I guess I could use something like `BufEnter`. Thanks anyway!

**Reply by wbthomason** [reply]
Hmm - what would usually happen automatically that should trigger the autosave plugin? It seems like (from what you've described) you'd always need to manually save, e.g. in your example case of loading on `:write`.

**Reply by ichigo-gyunyu** [reply]
Now that I think about it, yeah you are right. The autosave is triggered on `InsertLeave` and `TextChanged`. Even if I was able to lazy-load on `write` I'd have to manually save first...

**Reply by AckslD** [reply]
Just out of curiosity (since I recently went over my setup to lazy load bunch of things), what would be the benefit of lazy loading this? Since it seems you anyway always want it enabled?

**Reply by ichigo-gyunyu** [reply]
Lot of times I find myself just opening files to view their contents and not make any changes. I know we are dealing with is in the order of milliseconds, but I feel it's good practice to only load the plugins that are abosolutely necessary.

**Reply by wbthomason** [reply]
I think if you trigger on `TextChanged`, `TextChangedI`, and `TextChangedP`, you should be able to capture this behavior pretty well. If you just open the file and don't change anything, the autosave plugin doesn't load. As soon as you make a change in Normal or Insert mode, it gets loaded (and presumably saves as expected).

**Reply by AckslD** [reply]
That makes sense :)
